src/pipeline_migration/migrate.py

```diff
@@ -1,14 +1,356 @@
+import logging
+import os.path
+import re
+import subprocess as sp
+import tempfile
 
-import argparse
+from collections.abc import Generator
+from contextlib import contextmanager
+from dataclasses import dataclass, field
+from typing import Final, Any
 
+from pipeline_migration.utils import dump_yaml, load_yaml
+from pipeline_migration.registry import Container, Registry, ImageIndex
+from pipeline_migration.quay import QuayTagInfo, list_active_repo_tags
+from pipeline_migration.utils import is_true
+from pipeline_migration.types import FilePath
 
-def main():
-    parser = argparse.ArgumentParser(description="Pipeline migration tool for Konflux CI.")
+# TODO: once the build-and-push.sh is done, correct this name if necessary
+MIGRATION_ANNOTATION: Final[str] = "dev.konflux-ci.task.migration"
 
-    parser.add_argument("-f", required=True)
-    parser.add_argument("-t", required=True)
-    parser.add_argument("-p", required=True)
+TEKTON_KIND_PIPELINE: Final = "Pipeline"
+TEKTON_KIND_PIPELINE_RUN: Final = "PipelineRun"
+ANNOTATION_TRUTH_VALUE: Final = "true"
 
-    args = parser.parse_args()
+# Example:  0.1-18a61693389c6c912df587f31bc3b4cc53eb0d5b
+TASK_TAG_REGEXP: Final = r"^[0-9.]+-[0-9a-f]+$"
+DIGEST_REGEXP: Final = r"sha256:[0-9a-f]+"
 
-    print(args)
+logger = logging.getLogger("migrate")
+
+
+@dataclass
+class TaskBundleMigration:
+    # A complete image reference with both tag and digest
+    task_bundle: str
+    # Content of the script
+    migration_script: str
+
+
+@dataclass
+class TaskBundleUpgrade:
+    dep_name: str  # Renovate template field: depName
+    current_value: str  # Renovate template field: currentValue. It is the image tag.
+    current_digest: str  # Renovate template field: currentDigest
+    new_value: str  # Renovate template field: newValue. It is the image tag.
+    new_digest: str  # Renovate template field: newDigest
+
+    migrations: list[TaskBundleMigration] = field(default_factory=list)
+
+    @property
+    def current_bundle(self) -> str:
+        return f"{self.dep_name}:{self.current_value}@{self.current_digest}"
+
+    @property
+    def new_bundle(self) -> str:
+        return f"{self.dep_name}:{self.new_value}@{self.new_digest}"
+
+    @property
+    def comes_from_konflux(self) -> bool:
+        return self.dep_name.startswith("quay.io/konflux-ci/")
```

Comment:
For local testing, would it make sense to have an env var or cli arg for the allowed prefix? I've been changing this to `return True` for my tests, but that's not very convenient

Reply:
I can't figure out a better way to work around it at this moment. If doing that, it seems a little equivalent to removing the check `comes_from_konflux`.

Reply:
It would be a `comes_from_allowed_namespace` check.

Mintmaker could decide what is the allowed namespace by setting `--allowed-namespace=quay.io/konflux-ci` or a similar env var

Reply:
Added an env var `PMT_LOCAL_TEST`.